### PR TITLE
Feature/add category support

### PR DESCRIPTION
**Pull request by AtomicAlexD**
This pull request introduces significant updates to the `data_handler` module, the front-end logic for user and category management, and the associated test cases. Key changes include refactoring the database interaction methods, adding support for category management, enhancing the front-end to handle user and category selection, and expanding test coverage. Below are the most important changes grouped by theme:

### Database Interaction Enhancements:
* Refactored `read_database` to support parameterized queries and return results as a list. This method was moved and improved to handle SQL execution more robustly.
* Added `add_category` and `get_categories` methods to manage categories in the database, including support for optional budget values and parent categories.

### Front-End Updates:
* Updated the `home.html` template to include forms for selecting users, adding categories, and displaying categories. Added dynamic dropdowns and validation for user and category inputs.
* Enhanced the `home_page` function to fetch categories for the selected user and pass them to the template for rendering.
* Added new routes (`/select_user` and `/submit_category`) to handle user selection and category creation, with appropriate session management and error handling.

### Code Refactoring:
* Removed the old `read_database` method and replaced its usage with the new implementation. Simplified logging and error handling in `add_user_name`.

### Test Coverage Improvements:
* Added comprehensive test cases for new methods (`add_category`, `get_categories`) and enhanced existing tests to validate database interactions, including parameterized queries and edge cases.
* Introduced a test-specific database setup and teardown mechanism to ensure isolated and repeatable tests.

These changes collectively improve the maintainability, functionality, and testability of the application while introducing new features for category management.

### CODE REVIEW

File: src/data_handler.py [hunk 1]
```diff
@@ -184,6 +177,49 @@ def add_user_name(self, user_name: str) -> int:
...
+            categories = cursor.fetchall()
+        except Exception as e:
+            logger.error(f"Failed to fetch categories: {e}")
+            categories = []
+        finally:
+            try:
+                connection.close()
+            except Exception as e:
+                logger.error(f"failed to close connection: {e}")
+            return categories
```
Comment (Jake-Pullen) [inline src/data_handler.py:225]: self.read_database should be used to read from the database

File: src/front/home.py [hunk 2]
```diff
@@ -16,25 +15,119 @@ def home_page():
...
     users = db_call.get_users()
-    return render_template("home.html", users=users)
+    # Get user info directly from session
+    user_id = session.get("user_id") 
+    selected_user_name = session.get("user_name")
+    
+    # Debug log to see what's being passed to the template
+    logger.debug(f"User ID from session: {user_id}, User Name from session: {selected_user_name}")
+    
+    categories = db_call.get_categories(user_id) if user_id else []
```
Comment (Jake-Pullen) [inline src/front/home.py:25]: Categories here is used for 2 reasons right?
1 to select an optional parent category and 2 to show all the categories of a user. 
We have agreed a hard limit on a 3 tier hierarchy for parent / child category but currently this limit is not implemented. we might need to have a secondary get categories call with the limit enforced for the parent selection

File: src/front/templates/home.html [hunk 3]
```diff
@@ -1,25 +1,65 @@
...
     </form>
 </div>
 <div>
     <p>
-    {% for user in users %}
-- {{ user[0] }}<br>
+    <b>Your Categories:</b><br>
+    {% for category in categories %}
+        - {{ category[1] }}{% if category[3] %} (Sub of ID {{ category[3] }}){% endif %}{% if category[2] %} [Budget: {{ category[2] }}]{% endif %}<br>
     {% endfor %}
```
Comment (Jake-Pullen) [inline src/front/templates/home.html:57]: Can we display the categories in a something like a tree so its easily visible which ones are the parents and their children?

File: src/tests/test_data_handler.py [hunk 4]
```diff
@@ -1,59 +1,183 @@
...
+    assert isinstance(results, list)
+    assert len(results) > 0
+    assert results[0][1] == "param_test_user"
+
+
+def test_build_database(test_db_call):
+    """Test that the database builds correctly with all required tables"""
+    # Check if all tables have been created
+    tables_to_check = ["user", "account", "account_type", "party", "transaction_type", 
+                     "category", "transaction_log"]
```
Comment (Jake-Pullen) [inline src/tests/test_data_handler.py:86]: Can we load this from the schema yaml? just in case we add new tables later and forget to test them here

File: src/tests/test_data_handler.py [hunk 5]
```diff
@@ -1,59 +1,183 @@
...
+        assert len(results) == 1, f"Table '{table}' should exist"
 
 
 def test_add_user_name(test_db_call):
-    # Test adding a new user name
+    """Test adding a new user name"""
     user_name = "Test user"
     user_id = test_db_call.add_user_name(user_name)
-    assert user_id is not None
+    assert user_id > 0
```
Comment (Jake-Pullen) [inline src/tests/test_data_handler.py:97]: We should probably test we can add a new user name before adding users for other tests above?

File: src/tests/test_data_handler.py [hunk 6]
```diff
@@ -1,59 +1,183 @@
...
+    for user in users:
+        assert len(user) == 2
+        assert isinstance(user[0], int)  # ID
+        assert isinstance(user[1], str)  # Name
+
+
+def test_add_category_success(test_db_call):
+    # Add a user first, since category requires a user_id
+    user_id = test_db_call.add_user_name("CategoryTestUser")
+    assert user_id != 0
```
Comment (Jake-Pullen) [inline src/tests/test_data_handler.py:140]: This is testing adding a user, which we have already tested and is not in scope of this test

File: src/tests/test_data_handler.py [hunk 7]
```diff
@@ -1,59 +1,183 @@
...
+    assert isinstance(categories, list)
+    names = [cat[1] for cat in categories]
+    assert category_name in names
+    assert subcategory_name in names
+
+
+def test_add_category_invalid_user(test_db_call):
+    # Try to add a category with a non-existent user_id (should still succeed in SQLite unless foreign keys are enforced)
+    category_id = test_db_call.add_category(99999, "InvalidUserCategory")
+    assert isinstance(category_id, int)
```
Comment (Jake-Pullen) [inline src/tests/test_data_handler.py:183]: success or fail we always return an int (0 if fail)
so shouldn't the test to be assert category_id == 0 (as FK integrity breach)

**Review by Jake-Pullen — DISMISSED** [review]
Overall amazing progress. There is some comments to address and overall i'd like to see more logging and some minor formatting cleanup (suggest letting a linter change files on save, will fix whitespace on blank lines | trailing whitespace | blank lines count between functions)

File: src/front/home.py (hunk 2 above)
Comment (Jake-Pullen) [reply src/front/home.py:26]: Gonna add a new work item to handle categories hierarchy properly

File: src/front/templates/home.html (hunk 3 above)
Comment (Jake-Pullen) [reply src/front/templates/home.html:57]: Separate work item opened for this #18
